performance_test/src/communication_abstractions/ros2_communicator.hpp

```diff
@@ -0,0 +1,178 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef COMMUNICATION_ABSTRACTIONS__ROS2_COMMUNICATOR_HPP_
+#define COMMUNICATION_ABSTRACTIONS__ROS2_COMMUNICATOR_HPP_
+
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <memory>
+#include <atomic>
+
+#include "../experiment_configuration/topics.hpp"
+#include "../experiment_configuration/qos_abstraction.hpp"
+
+#include "communicator.hpp"
+#include "resource_manager.hpp"
+
+namespace performance_test
+{
+
+/// Translates abstract QOS settings to specific QOS settings for ROS 2.
+class ROS2QOSAdapter
+{
+public:
+  /**
+   * \brief The constructor which will save the provided abstract QOS.
+   * \param qos The abstract QOS to derive the settings from.
+   */
+  explicit ROS2QOSAdapter(const QOSAbstraction qos)
+  : m_qos(qos)
+  {}
+  /// Gets a ROS 2 QOS profile derived from the stored abstract QOS.
+  inline rclcpp::QoS get() const
+  {
+    rclcpp::QoS ros_qos(m_qos.history_depth);
+
+    if (m_qos.reliability == QOSAbstraction::Reliability::BEST_EFFORT) {
+      ros_qos.best_effort();
+    } else if (m_qos.reliability == QOSAbstraction::Reliability::RELIABLE) {
+      ros_qos.reliable();
+    } else {
+      throw std::runtime_error("Unsupported QOS!");
+    }
+
+    if (m_qos.durability == QOSAbstraction::Durability::VOLATILE) {
+      ros_qos.durability_volatile();
+    } else if (m_qos.durability == QOSAbstraction::Durability::TRANSIENT_LOCAL) {
+      ros_qos.transient_local();
+    } else {
+      throw std::runtime_error("Unsupported QOS!");
+    }
+
+
+    if (m_qos.history_kind == QOSAbstraction::HistoryKind::KEEP_ALL) {
+      ros_qos.keep_all();
+    } else if (m_qos.history_kind == QOSAbstraction::HistoryKind::KEEP_LAST) {
+      ros_qos.keep_last(m_qos.history_depth);
+    } else {
+      throw std::runtime_error("Unsupported QOS!");
+    }
+
+    return ros_qos;
+  }
+
+private:
+  const QOSAbstraction m_qos;
+};
+
+/// Communication plugin interface for ROS 2 for the subscription side.
+template<class Topic>
+class ROS2Communicator : public Communicator
+{
+public:
+  /// The data type to publish and subscribe to.
+  using DataType = typename Topic::RosType;
+
+  /// Constructor which takes a reference \param lock to the lock to use.
+  explicit ROS2Communicator(SpinLock & lock)
+  : Communicator(lock),
+    m_node(ResourceManager::get().ros2_node())
+  {
+    m_ROS2QOSAdapter = std::make_shared<rclcpp::QoS>(ROS2QOSAdapter(m_ec.qos()).get());
```

Comment:
Why is that a pointer?

Reply:
Yes doesn't need to be a pointer. I can change that,